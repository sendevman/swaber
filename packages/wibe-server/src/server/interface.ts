import type { WobeResponse } from 'wobe'
import type { User } from '../generated/wibe'
import type { DatabaseController } from '../database'
import type { WibeAppTypes } from '.'

export interface Context {
	response?: WobeResponse
	user?: User | null
	sessionId?: string
	isRoot: boolean
	databaseController: DatabaseController<WibeAppTypes>
}
