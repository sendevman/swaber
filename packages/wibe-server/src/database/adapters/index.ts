export interface DatabaseAdapter {
	connect(): Promise<void>
	disconnect(): Promise<void>
	insert(): Promise<void>
}
