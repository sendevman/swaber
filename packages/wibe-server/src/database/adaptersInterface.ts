import { Document, WithId } from 'mongodb'

export interface AdapterOptions {
	databaseUrl: string
	databaseName: string
}

// TODO : Type the return of the functions
export interface DatabaseAdapter {
	connect(): Promise<any>
	createClass(className: string): Promise<any>
	getObject(params: {
		className: string
		id: string
		fields: Record<string, any>
	}): Promise<WithId<Document>>
	insertObject(params: {
		className: string
		data: Record<string, any>
	}): Promise<any>
	updateObject(params: {
		className: string
		id: string
		data: Record<string, any>
	}): Promise<any>
}
