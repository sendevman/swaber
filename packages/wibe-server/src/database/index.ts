export enum DatabaseEnum {
	Mongo = 'mongo',
}

export interface DatabaseConfig {
	type: DatabaseEnum
	url: string
}
