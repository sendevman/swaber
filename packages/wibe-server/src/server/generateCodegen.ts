import type {
	ClassInterface,
	EnumInterface,
	ScalarInterface,
	SchemaInterface,
} from '../schema'

export const generateWibeFile = ({
	scalars,
	enums,
	classes,
}: {
	enums?: EnumInterface[]
	scalars?: ScalarInterface[]
	classes: ClassInterface[]
}) => {
	// Scalars
	const listOfScalars = scalars?.map((scalar) => `"${scalar.name}"`) || []
	const wibeScalarType = `export type WibeSchemaScalars = ${listOfScalars.join(
		' | ',
	)}`

	// Enums
	const wibeEnumsGlobalTypes =
		enums?.map((wibeEnum) => `"${wibeEnum.name}"`) || []

	const wibeEnumsGlobalTypesString = `export type WibeSchemaEnums = ${wibeEnumsGlobalTypes.join(
		' | ',
	)}`

	// Classes
	const allNames = classes
		.map((schema) => `${schema.name}: ${schema.name}`)
		.filter((schema) => schema)

	const globalWibeTypeString = `export type WibeSchemaTypes = {\n\t${allNames.join(
		',\n\t',
	)}\n}`

	return `${wibeScalarType}\n\n${wibeEnumsGlobalTypesString}\n\n${globalWibeTypeString}`
}

export const generateCodegen = async ({
	schema,
	graphqlSchema,
	path,
}: {
	schema: SchemaInterface
	graphqlSchema: string
	path: string
}) => {
	Bun.write(
		`${path}/wibe.ts`,
		`${generateWibeFile({
			scalars: schema.scalars,
			enums: schema.enums,
			classes: schema.classes,
		})}`,
	)

	Bun.write(`${path}/schema.graphql`, graphqlSchema)
}
