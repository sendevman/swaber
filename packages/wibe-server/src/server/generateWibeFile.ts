import type { ClassInterface, EnumInterface, ScalarInterface } from '../schema'

export const generateWibeFile = ({
	scalars,
	enums,
	schemas,
}: {
	enums?: EnumInterface[]
	scalars?: ScalarInterface[]
	schemas: ClassInterface[]
}) => {
	// Scalars
	const listOfScalars = scalars?.map((scalar) => `"${scalar.name}"`) || []
	const wibeScalarType = `export type WibeSchemaScalars = ${listOfScalars.join(
		' | ',
	)}`

	// Enums
	const wibeEnumsGlobalTypes =
		enums?.map((wibeEnum) => `"${wibeEnum.name}"`) || []

	const wibeEnumsGlobalTypesString = `export type WibeSchemaEnums = ${wibeEnumsGlobalTypes.join(
		' | ',
	)}`

	// Types
	const allNames = schemas
		.map((schema) => `${schema.name}: ${schema.name}`)
		.filter((schema) => schema)

	const globalWibeTypeString = `export type WibeSchemaTypes = {\n\t${allNames.join(
		',\n\t',
	)}\n}`

	return `${wibeScalarType}\n\n${wibeEnumsGlobalTypesString}\n\n${globalWibeTypeString}`
}
