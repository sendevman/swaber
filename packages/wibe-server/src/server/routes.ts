import { Context } from 'elysia'
import { WibeApp } from '.'
import { GoogleProvider } from '../authentication/providers/google'

export const googleAuthHandler = async (context: Context) => {
	const code = context.query.code
	const { authentication: authenticationConfig } = WibeApp.config

	if (!authenticationConfig)
		throw new Error('Authentication config not found')

	// 296431040556-4jh84e5s264rmrgnh8bmegb0kl550teg.apps.googleusercontent.com
	const clientId = authenticationConfig.providers.GOOGLE.clientId
	// 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
	const clientSecret = authenticationConfig.providers.GOOGLE.clientSecret

	if (!code || !clientId)
		throw new Error('Authentication : Google client id or secret not found')

	const googleProvider = new GoogleProvider(clientId, clientSecret)

	try {
		const { accessToken, refreshToken } =
			await googleProvider.validateTokenFromAuthorizationCode(code)

		// Create cookie for access and refresh token
		context.cookie.accessToken.add({
			value: accessToken,
			httpOnly: true,
			path: '/',
			expires: new Date(Date.now() + 1000 * 60 * 60 * 24 * 7),
		})

		context.cookie.refreshToken.add({
			value: refreshToken,
			httpOnly: true,
			path: '/',
			expires: new Date(Date.now() + 1000 * 60 * 60 * 24 * 7),
		})

		context.set.redirect = 'http://localhost:5173'
	} catch (e) {
		console.error(e)
		context.set.redirect = authenticationConfig?.failureRedirectPath
	}
}
