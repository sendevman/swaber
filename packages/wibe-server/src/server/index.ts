import { Elysia } from 'elysia'
import { apollo, gql } from '@elysiajs/apollo'
import { DatabaseConfig, DatabaseEnum } from '../database'
import { SchemaInterface } from '../schema/interface'
import { Schema } from '../schema'
import { DatabaseController } from '../database/controllers/DatabaseController'
import { SchemaRouterController } from '../schema/controllers/SchemaRouterController'
import { GraphQLSchemaAdapter } from '../schema/adapters'
import { makeSchema } from 'nexus'
import { MongoAdapter } from '../database/adapters/MongoAdapter'

interface WibeConfig {
	port: number
	schema: SchemaInterface[]
	database: DatabaseConfig
}

export class WibeApp {
	private config: WibeConfig
	private server: Elysia

	constructor(config: WibeConfig) {
		this.config = config

		this.server = new Elysia().get(
			'/health',
			(context) => (context.set.status = 200),
		)

		const databaseAdapter = new MongoAdapter({
			databaseName: this.config.database.name,
			databaseUrl: this.config.database.url,
		})

		const databaseController = new DatabaseController(databaseAdapter)

		const schemas = this.config.schema.map(
			(schema) =>
				new Schema({
					name: schema.name,
					fields: schema.fields,
					databaseController,
				}),
		)

		const schemaRouterAdapter = new GraphQLSchemaAdapter(schemas)

		const schemaRouterController = new SchemaRouterController(
			schemaRouterAdapter,
		)
		const types = schemaRouterController.createSchema()

		const graphqlSchema = makeSchema({ types })

		this.server.use(apollo({ schema: graphqlSchema }))

		this.server.listen(this.config.port)
	}

	async close() {
		return this.server.stop()
	}
}

const wibe = new WibeApp({
	database: {
		type: DatabaseEnum.Mongo,
		url: 'mongodb://localhost:27017',
		name: 'wibe',
	},
	port: 3000,
	schema: [
		{
			name: 'user',
			fields: {
				name: { type: 'String' },
				age: { type: 'Int' },
			},
		},
	],
})
