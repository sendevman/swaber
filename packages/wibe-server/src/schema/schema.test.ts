import { beforeAll, describe, expect, it } from 'bun:test'
import { Schema } from './Schema'
import { DatabaseEnum } from '../database'
import { WibeApp } from '../server'

describe('Schema', () => {
    let schema: Schema

    beforeAll(async () => {
        schema = new Schema({
            class: [
                {
                    name: 'Collection1',
                    fields: {
                        name: {
                            type: 'String',
                            defaultValue: 'Lucas',
                        },
                        age: { type: 'Int', defaultValue: 23 },
                        isReady: {
                            type: 'Boolean',
                            defaultValue: true,
                        },
                    },
                },
            ],
            scalars: [
                {
                    name: 'Phone',
                    description: 'Phone custom scalar type',
                },
                {
                    name: 'Email',
                    description: 'Email custom scalar type',
                },
            ],
            enums: [
                {
                    name: 'Role',
                    values: {
                        Admin: 'admin',
                        Member: 'member',
                    },
                },
            ],
        })
    })

    it('should write schema types in output file', () => {
        const output = schema.getTypesFromSchema()

        expect(output).toMatchSnapshot()
    })

    it("should throw an error if the scalar doesn't exist", async () => {
        const wibe = new WibeApp({
            database: {
                type: DatabaseEnum.Mongo,
                url: 'mongodb://127.0.0.1:27045',
                name: 'Wibe',
            },
            port: 3000,
            schema: {
                class: [
                    {
                        name: 'User',
                        fields: {
                            name: {
                                // @ts-ignore
                                type: 'ScalarThatDoesntExist',
                            },
                        },
                    },
                ],
            },
        })

        expect(wibe.start()).rejects.toThrow(
            'ScalarThatDoesntExist not exist in schema',
        )
    })
})
