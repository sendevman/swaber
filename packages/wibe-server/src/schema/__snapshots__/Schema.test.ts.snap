// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema should write schema types in output file 1`] = `
"export type Collection1 = {
  id: string,
  name: string,
  age: number,
  isReady: boolean
}

export type WibeSchemaScalars = "Phone" | "Email"

export enum Role {
	Admin = 'admin',
	Member = 'member',
}

export type WibeSchemaEnums = "Role"

export type WibeSchemaTypes = {
  Collection1: Collection1
}"
`;
